### ISSUE TITLE
Schema for Model Metadata

### ISSUE THREAD

**Issue by isaaccorley** [microsoft/torchgeo#2880]
### Summary

We currently use the Torchvision Weights API which allows for a metadata dict to be attached to a model's weights to couple some necessary fields to the model e.g. in_channels, num_classes, license, model type, etc.

```python
    LANDSAT_TM_TOA_SIMCLR = Weights(
        url='https://hf.co/torchgeo/ssl4eo_landsat/resolve/1c88bb51b6e17a21dde5230738fa38b74bd74f76/resnet18_landsat_tm_toa_simclr-d2d38ace.pth',
        transforms=_ssl4eo_l_transforms,
        meta={
            'dataset': 'SSL4EO-L',
            'in_chans': 7,
            'model': 'resnet18',
            'publication': 'https://arxiv.org/abs/2306.09424',
            'repo': 'https://github.com/microsoft/torchgeo',
            'ssl_method': 'simclr',
            'bands': _landsat_tm_toa_bands,
        },
    )
```

However, we currently have no way to enforce that a minimum set of fields are filled out. It would be ideal to start doing this so that when anyone lists all model metadata, the naming and fields would all follow the same structure and typing.

The [stac_model](https://github.com/stac-extensions/mlm) package has pydantic classes that can be used to validate this. For example given a  model's metadata we can validate that the necessary minimal required fields are filled out (additional fields are fine to add and do not break the schema).

Note: I'm using metadata as a yaml but this can be a dict or a json file (doesn't matter as long as it's passed through the pydantic class for validation)

```yaml
name: Unet
architecture: Unet
artifact_type: torch.export.save
framework: torch
framework_version: 2.7.0
accelerator: cuda
total_parameters: 1234567
tasks: [semantic-segmentation]
input:
  - name: Imagery
    bands: [red, blue, green]
    input:
      shape: [-1, 3, 512, 512]
      dim_order: [batch, channel, height, width]
      data_type: float32
      crs: EPSG:4326
      res: 0.1
    pre_processing_function:
      format: pt2-exported-program # indicates that expression is interpreted as a key referencing exported pytorch file that can be loaded to run the function
      expression: transforms
output:
  - name: segmentation-output
    tasks: ["semantic-segmentation"]
    result:
      shape: [-1, 6, 512, 512]
      dim_order: [batch, classes, height, width]
      data_type: float32
    classification:classes:
      - value: 0
        name: Class 0
      - value: 1
        name: Class 1
      - value: 2
        name: Class 2
      - value: 3
        name: Class 3
      - value: 4
        name: Class 4
      - value: 5
        name: Class 5
```

```python
import yaml
from stac_model.schema import MLModelProperties

with open("path/to/metadata/yaml", "r") as f:
    metadata = yaml.safe_load(f)

MLModelProperties(**metadata)  # this raises errors if required fields are missing or do not have the correct types
```

Some benefits of standardization to a schema:
- **Adopting a standard that is gaining traction in the community**: several organizations already utilize the MLM STAC Extension for their models (e.g. Regrow, Taco Foundation, etc.)
- **Inference automation**: If we can expect that some minimal set of fields should come packaged with each model or in a model's checkpoint file -- inference code can be better automated to load a model and use it for some downstream task

Suggestion (I can take on these tasks btw):
- We should add `stac_model` as a dependency for performing this metadata validation
- We should backfill and validate all current model metadata
- We should enforce when a new model is added to TorchGeo that the minimal set of fields are filled out
- We should add a feature to for packaging model metadata to be exported with model checkpoints trained by our existing trainers

CC: @rbavery @ ljstrnadiii @adamjstewart @calebrob6

**Comment by adamjstewart** [comment]
My initial reaction:

* Standardization is good, but this could also be enforced by a custom class or `TypedDict`
* I'm always hesitant about adding new deps, but I've heard good things about pydantic
* I'm not married to following the torchvision API, I would much rather focus on the timm registry

This should not be taken as me not approving of this idea, just trying to play devil's advocate. I'm by no means up-to-date on the STAC ecosystem. I want to better integrate STAC into TorchGeo, but every time I try, it feels daunting (too many standards, too many libraries).

I guess my first question is, if we follow your suggestion [here](https://github.com/microsoft/torchgeo/discussions/2838#discussioncomment-13533364), what all changes? How does timm typically represent weights? Do they use a custom class, or dict, or something else? Are there already standards there? Do they conflict with anything in MLM? I would much rather make the torchvision -> timm API migration first, then take another look at this suggestion.

**Comment by fmigneault** [comment]
MLM does not impose constraints between `torhcvision`, `timm`, or any other weights representation. In fact, it is voluntarily abstract about these definitions to allow this flexibility in picking whichever one is more appropriate for the model/library using it,  even beyond `torch` or deep-learning in general (eg.: describing `scikit-learn` statistical models). Using MLM would allow this transition to `timm` with a minimal adjustment of the reference where the weights are actually downloaded from. 

The focus of MLM is much more oriented toward the metadata that can help understand what the model expects and produces as input/output, the relevant data preparation, runtime constraints, scientific references, or reproducibility considerations required to properly apply the model. It also focuses on the specific domain considerations of spatio-temporal, earth observation, and climate data, combined to the more "traditional" computer vision and machine learning aspects involved with such models.

An advantage of MLM over a custom Python definition would be that it relies on the STAC ecosystem, which means that the connection between the data retrieved and described using STAC can be more easily mapped to the relevant models and how to make them work together. There is also a lot of effort put into describing [MLM Best Practices](https://github.com/stac-extensions/mlm/blob/main/best-practices.md) to unify the comprehension of multiple STAC extensions found in the wild. We would be happy to put forward even more details in best practices specifically for `torchgeo` if the alignment is more straightforward (eg.: cross-referencing specific MLM metadata that the library provides or recommended methods to describe models directly offered by `torchgeo`).

**Comment by adamjstewart** [comment]
I'm curious how it handles multi-modal foundation models like Copernicus-FM that are pretrained on Sentinel-1/2/3/5P + Copernicus DEM. Maybe we can start with a single complex model just to see what it looks like.

**Comment by fmigneault** [comment]
For multi-modal data, there shouldn't be any issues. The input/output definitions can provide multiple domains and their respective bands per corresponding satellite imagery inputs, or output climate variables/land-cover pixel values, etc., as applicable. For the model itself, it would have to be described using its definition script (eg: https://github.com/zhu-xlab/Copernicus-FM/blob/main/Copernicus-FM/src/model_vit.py) and the relevant weights. The data preparation steps (https://github.com/zhu-xlab/Copernicus-FM/tree/main/Copernicus-Pretrain) could also be cross-referenced by MLM.

**Comment by kshitijrajsharma** [comment]
Awesome , I am curious to see how this issue progresses forward , related thread : https://github.com/torchgeo/torchgeo/issues/2732#issuecomment-2810981505

**Comment by isaaccorley** [comment]
> Awesome , I am curious to see how this issue progresses forward , related thread : https://github.com/torchgeo/torchgeo/issues/2732#issuecomment-2810981505 

Hi @kshitijrajsharma! We've made some progress for packaging an exported model, transforms, and its metadata into a single PT2 archive in this PR https://github.com/stac-extensions/mlm/pull/97. You can see an example of how to use it below.

```python
import torch
import torchvision.transforms.v2 as T
from torchgeo.models import Unet_Weights, unet
from stac_model.torch.export import save

weights = Unet_Weights.SENTINEL2_3CLASS_FTW
transforms = torch.nn.Sequential(
  T.Resize((256, 256)),
  T.Normalize(mean=[0.0], std=[3000.0])
)
model = unet(weights=weights)

save(
    output_file="ftw.pt2",
    model=model,  # Must be an nn.Module
    transforms=transforms,  # Must be an nn.Module
    metadata_path="metadata.yaml",  # Can be a metadata yaml or stac_model.schema.MLModelProperties object
    input_shape=[-1, 8, -1, -1],  # -1 indicates a dynamic shaped dimension
    device="cpu",
    dtype=torch.float32,
    aoti_compile_and_package=False,  # True for AOTInductor compile otherwise use torch.export
)
```

**Comment by kshitijrajsharma** [comment]
Thanks @isaaccorley  , Thanks a million for working on this !  Looking forward to test it out and mostly try to make it available with the work we are doing on  https://fair.hotosm.org/  , will provide more update on https://github.com/hotosm/fAIr/issues/453 !

**Comment by fmigneault** [comment]
Extra example (https://github.com/stac-extensions/mlm/blob/main/README_STAC_MODEL.md#creating-a-stac-item-from-a-pytorch-model) also show methods that attempt auto-detecting some of the model's metadata. 

We still need to work on bridging the parts into a full pipeline, but the initial components are in place.
```
model -> from_torch(model) -> MLM -> save(model, MLM) -> ExportProgram -> package() -> archive PT2
```

**Comment by isaaccorley** [comment]
> Thanks [@isaaccorley](https://github.com/isaaccorley) , Thanks a million for working on this ! Looking forward to test it out and mostly try to make it available with the work we are doing on [fair.hotosm.org](https://fair.hotosm.org/) , will provide more update on [hotosm/fAIr#453](https://github.com/hotosm/fAIr/issues/453) !

Thanks for being excited about it! We would love users to test it out!

After exporting, you can load the model, transforms, metadata from the PT2 archive and run it using the following:

```python
import torch
import yaml
from torch.export.pt2_archive._package import load_pt2

model = pt2.exported_programs["model"].module()
transforms =  pt2.exported_programs["transforms"].module()
metadata = yaml.safe_load(pt2.extra_files["mlm-metadata"])

with torch.inference_mode():
   x = torch.randn(1, in_channels, h, w, device=device, dtype=dtype)
   preds = model(transforms(x))
```

**Comment by fmigneault** [comment]
Hi @isaaccorley 
I was wondering how is the status of the initiative going, from torchgeo and OSM side?
Are there any remaining blockers from the STAC MLM end?

**Comment by kshitijrajsharma** [comment]
Hi @fmigneault  Yes , from our side: We  kept this  in our road map for early 2026 , You will hear back from me with progress after the project kicks off ! 

**Update :** 
Here is one of the example i built that we are testing on staging env for one sample project form torchgeo : https://github.com/hotosm/fAIr-models/blob/master/models/unet_segmentation/stac-item.json

**Comment by isaaccorley** [comment]
> Hi @isaaccorley 
> I was wondering how is the status of the initiative going, from torchgeo and OSM side?
> Are there any remaining blockers from the STAC MLM end?

Hey Francis -- the only blocker right now is we need to merge the PR that converts our model inference transforms to use torchvision instead of kornia. Kornia transforms can't be exported which is unfortunate but necessary to be able to export all models and transforms. Also cc: @adamjstewart

**Comment by rbavery** [comment]
I think the above is now complete. Just flagging that Isaac is on family leave and can get back to this in the new year when he returns.

**Comment by robmarkcole** [comment]
Any plans to kick this off again @isaaccorley ?
